src/sandstorm/indexer.capnp

```diff
@@ -0,0 +1,139 @@
+  # with that event separately specifically for people who can see said event. Hence, if your app
+  # has content that is visible to some users but not others, and that needs to be indexed, it
+  # should be sure to log appropriate activity events.
+
+  indexAll @0 (indexer :GrainIndexer(Metadata));
+  # Asks the app to iterate through the grain's entire text contents, pushing it all to the given
+  # indexer capability.
+
+  indexPaths @1 (paths :List(Text), indexer :GrainIndexer(Metadata));
+  # Index a specific list of paths. If a path in the list doesn't exist, calls `indexer.index()`
+  # with `content` = null for that path.
+}
+
+interface GrainIndexer(Metadata) {
+  # Capability used to index the content of a grain.
+  #
+  # This is a one-way capability. Although GrainIndexer is implemented by the indexer and is called
+  # by arbitrary content grains, Sandstorm ensures that information cannot leak from the indexer to
+  # the content grains by implementing a one-way message queue in between. When the content grain
+  # calls index(), Sandstorm places the data in a queue and returns immediately. Later on,
+  # Sandstorm takes calls from the queue and actually delivers them to the indexer. This means
+  # that not only is the indexer unable to return data to the caller, but the caller cannot find
+  # out how long it takes to process the call nor if it threw an exception.
+
+  index @0 (path :Text, content :IndexableContent(Metadata));
+  # Add content of the given path to the index.
+  #
+  # A null value for `content` indicates that the path doesn't exist.
+  #
+  # TODO(someday): When Cap'n Proto supports bulk-transfer methods with flow control, mark this
+  #   method as such. This will cause the Cap'n Proto implementation to pretend the method has
+  #   completed (resolving the promise) as soon as it's appropriate for the caller to make another
+  #   call. For now, apps should pretend this is already in place, and make only one call at a
+  #   time -- the current performance penalty in doing so probably isn't a big deal.
+}
+
+struct IndexableContent(Metadata) {
+  body @0 :Text;
+  # Freeform natural-language body text. Will be tokenized by the search indexer.
+
+  links @1 :List(Grain.UiView);
+  # Other grains that can be accessed through this content.
+
+  threadPath @2 :Text;
+  # Optional path of thread of which this item is a part, if any. See `ActivityEvent.thread`.
+
+  metadata @3 :Metadata;
+  # Metadata for search operators, e.g. "subject:", "author:", etc. Each app can define its own
+  # metadata format.
+  #
+  # TODO(someday): Spec out how this works. Not needed for MVP of search.
+}
+
+# ========================================================================================
+# Indexer implementation
+
+interface IndexerSession {
+  # A session type specifically implemented by the indexer app. The app's UiView's newSession()
```

Comment:
How many indexer app grains are there? Just one? One per user? Maybe the are created on demand when the user manually sets up some grains as indexable?